File: video/src/main/java/com/goog/video/gl/ESurfaceTexture.java
package com.goog.video.gl;

import android.graphics.SurfaceTexture;

import com.goog.video.filter.GlPreviewFilter;

class ESurfaceTexture implements SurfaceTexture.OnFrameAvailableListener {

    private final SurfaceTexture surfaceTexture;

    private SurfaceTexture.OnFrameAvailableListener onFrameAvailableListener;

    public ESurfaceTexture(final int texName) {
        surfaceTexture = new SurfaceTexture(texName);
        surfaceTexture.setOnFrameAvailableListener(this);
    }

    void setOnFrameAvailableListener(final SurfaceTexture.OnFrameAvailableListener l) {
        onFrameAvailableListener = l;
    }

    int getTextureTarget() {
        return GlPreviewFilter.GL_TEXTURE_EXTERNAL_OES;
    }

    void updateTexImage() {
        surfaceTexture.updateTexImage();
    }

    void getTransformMatrix(final float[] mtx) {
        surfaceTexture.getTransformMatrix(mtx);
    }

    SurfaceTexture getSurfaceTexture() {
        return surfaceTexture;
    }

    public void onFrameAvailable(final SurfaceTexture surfaceTexture) {
        if (onFrameAvailableListener != null) {
            onFrameAvailableListener.onFrameAvailable(this.surfaceTexture);
        }
    }

    public void release() {
        surfaceTexture.release();
    }

}

File: video/src/main/java/com/goog/video/gl/ESurfaceTexture.kt
package com.goog.video.gl

import android.graphics.SurfaceTexture
import android.graphics.SurfaceTexture.OnFrameAvailableListener
import android.view.Surface
import com.goog.video.filter.core.GLPreviewFilter

internal class ESurfaceTexture(texName: Int) : OnFrameAvailableListener {
    private val texture: SurfaceTexture = SurfaceTexture(texName)

    private var frameAvailableListener: OnFrameAvailableListener? = null

    init {
        texture.setOnFrameAvailableListener(this)
    }

    fun setFrameAvailableListener(l: OnFrameAvailableListener?) {
        frameAvailableListener = l
    }

    val textureTarget: Int
        get() = GLPreviewFilter.GL_TEXTURE_EXTERNAL_OES

    fun updateTexImage() {
        texture.updateTexImage()
    }

    fun getTransformMatrix(mtx: FloatArray?) {
        texture.getTransformMatrix(mtx)
    }

    override fun onFrameAvailable(surfaceTexture: SurfaceTexture) {
        frameAvailableListener?.onFrameAvailable(this.texture)
    }

    fun release() {
        texture.release()
    }

    fun obtainSurface(): Surface {
        return Surface(texture)
    }
}
